scripts/gpu_support/nvidia/link_nvidia_host_libraries.sh

```diff
@@ -1,144 +1,378 @@
+        rm -f $nvidia_smi_out
+    else
+        fatal_error "nvidia-smi command failed, see output in $nvidia_smi_out"
+    fi
 else
-    error="$command_name not found in PATH or only found in paths starting with $exclude_prefix."
-    fatal_error "$error"
+    fatal_error "nvidia-smi command not found"
+    exit 2
 fi
 
-# Make sure EESSI is initialised (doesn't matter what version)
-check_eessi_initialised
+# Gather any CUDA related driver libraries from the host
+# - First let's see what driver libraries are there
+# - then extract the ones we need for CUDA
 
-# Find the CUDA version of the host CUDA drivers
-# (making sure that this can still work inside prefix environment inside a container)
-export LD_LIBRARY_PATH=/.singularity.d/libs:$LD_LIBRARY_PATH
-nvidia_smi_command="nvidia-smi --query-gpu=driver_version --format=csv,noheader"
-if $nvidia_smi_command > /dev/null; then
-  host_driver_version=$($nvidia_smi_command | tail -n1)
-  echo_green "Found NVIDIA GPU driver version ${host_driver_version}"
-  # If the first worked, this should work too
-  host_cuda_version=$(nvidia-smi  -q --display=COMPUTE | grep CUDA | awk 'NF>1{print $NF}')
-  echo_green "Found host CUDA version ${host_cuda_version}"
+# Find the host ldconfig
+host_ldconfig=$(get_host_ldconfig)
+# Gather libraries on the host (_must_ be host ldconfig)
+host_libraries=$($host_ldconfig -p | awk '{print $NF}')
+singularity_libs=$(ls /.singularity.d/libs/* 2>/dev/null)
+
+# Now gather the list of possible CUDA libraries and make them into an array
+cuda_candidate_libraries=($(get_nvlib_list "${LIBS_LIST}"))
+# Check if the function returned an error (e.g., curl failed)
+if [ $? -ne 0 ]; then
+    echo "Using default list of libraries"
 else
-  error="Failed to successfully execute\n  $nvidia_smi_command\n"
-  fatal_error "$error"
+    echo "Using downloaded list of libraries"
 fi
 
-# Let's make sure the driver libraries are not already in place
+# Filter the host libraries to find the CUDA libaries locations
+# Initialize an array to hold the matched libraries
+matched_libraries=()
+
+# Process each library and check for matches in libs.txt
+for library in "${cuda_candidate_libraries[@]}"; do
+    # Search for the library in libs.txt and add it to the matched_libraries array
+    matched=$(echo "$host_libraries $singularity_libs" | grep "$library")
+    if [ -n "$matched" ]; then
+        matched_libraries+=( $matched )  # Add matched library to the array
+    fi
+done
+
+# Output the number of matched libraries
+echo "Matched ${#matched_libraries[@]} CUDA Libraries"
+
+# LD_PRELOAD Mode
+if [ "$LD_PRELOAD_MODE" -eq 1 ]; then
+    echo
+    echo_yellow "When attempting to use LD_PRELOAD we exclude anything related to graphics"
+
+    # Filter out all symlinks and libraries that have missing library dependencies under EESSI
+    filtered_libraries=()
+    for library in "${matched_libraries[@]}"; do
+        if [ ! -L "$library" ]; then
```

Comment:
This is too aggressive, instead we should just resolve the symlink and remove duplicate entries